syntax = "proto3";

package python_example.stock;
option java_package = "stock.generated";
option java_multiple_files = true;

message GetProductRequest {
}

message CreateProductRequest {
  int32 price = 1;
}

message DeleteProductRequest {
}

message AddProductStockRequest {
  int32 amount = 1;
}

message SubtractProductStockRequest {
  int32 amount = 1;
}

message StockRequest {
  string request_id = 1;
  string product_id = 2;

  oneof message {
    GetProductRequest get_product = 3;
    CreateProductRequest create_product = 4;
    DeleteProductRequest delete_product = 5;
    AddProductStockRequest add_product_stock = 6;
    SubtractProductStockRequest subtract_product_stock = 7;
  }
}

message StockResponse {
  string request_id = 1;
  string product_id = 2;
  int32 status_code = 3;
  
  oneof payload {
    string string = 4;
    ProductState state = 5;
  }
}

message ProductState {
  bool deleted = 1;
  int32 stock = 2;
  int32 price = 3;
}